optimum/habana/transformers/generation/utils.py

```diff
@@ -2873,10 +2873,18 @@ def _sample(
         if batch_size > 1 and has_eos_stopping_criteria:
             eos_token_id = generation_config.eos_token_id
             # Find the positions of the first eos_token_id in each sequence
-            eos_positions = (
-                torch.isin(input_ids[:, start_token_idx:], torch.tensor(eos_token_id)).int().argmax(dim=1)
-                + start_token_idx
-            )
+            _, seq_len = input_ids.shape
+            if start_token_idx >= seq_len:
+                # All slices are empty, so no EOS can be found
+                eos_positions = torch.full((batch_size,), seq_len, dtype=torch.long, device=input_ids.device)
+            else:
+                mask_ = torch.isin(input_ids[:, start_token_idx:], torch.tensor(eos_token_id, device=input_ids.device))
+                # If mask is all False for a row, argmax returns 0, which is incorrect.
+                # So, set eos_positions to seq_len if no EOS is found.
+                found = mask_.any(dim=1)
+                first_eos = mask_.int().argmax(dim=1) + start_token_idx
+                eos_positions = torch.full((batch_size,), seq_len, dtype=torch.long, device=input_ids.device)
+                eos_positions[found] = first_eos[found]
```

Comment:
```suggestion
            eos_positions = torch.full((batch_size,), seq_len, dtype=torch.long, device=input_ids.device)
            if start_token_idx < seq_len:
                # look for the first EOS in every row (starting at start_token_idx)
                slice_ = input_ids[:, start_token_idx:]
                eos_in_slice = slice_ == eos_token_id

                # argmax returns the FIRST 1 in every row; all-0 rows → 0
                first_rel = eos_in_slice.int().argmax(dim=1)
                # rows that actually contain an EOS anywhere (including index 0)
                has_eos = eos_in_slice.any(dim=1)

                eos_positions[has_eos] = first_rel[has_eos] + start_token_idx
```


@12010486 can you test the suggestion. Using `torch.isin` here for a single `eos_token_id` is not correct as it is designed to test membership against an arbitrary set of values. Internally it builds a hash-table (or sorts+binary-searches) so that every element of input_ids can be compared against every element in the test set.

Reply:
@yafshar eos can be more than a single value, at least when you use this code from run_lm_eval.py (it was a list of three for llama-3.2-1b-Instruct). Let me check with the example I shared instead

Reply:
Regarding the  _I can change the code to include the special case of single eos_token_id_ this would not be my preference, it seems this way is more generic and the logic holds because eos > single-element. What do you think?

Reply:
I'm okay with `torch.isin` if there is no perf regression.
I agree with @yafshar's suggestion to remove the duplicate code.

Reply:
Just to confirm that I'm on it at the moment. I could see that the major difference in performance is actually wrt the old/legacy code. Around 27-33x slower when running on lazy (yaser's suggestion at 27, mine 33), and 8-10x when running on eager (original PR proposal is faster this time), so I'm looking for something with better perf

Reply:
Done! Whenever you have time for a final check, the proposed is the fastest solution I've found. I can share more details if needed